File: core/src/com/alta_v2/rendering/component/box/BoxComponent.java
package com.alta_v2.rendering.component.box;

import com.alta_v2.rendering.Renderer;
import com.alta_v2.rendering.ScreenState;
import com.alta_v2.rendering.utils.GradientResource;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Rectangle;
import com.google.inject.assistedinject.Assisted;
import com.google.inject.assistedinject.AssistedInject;

import java.util.List;

public class BoxComponent implements Renderer {

    private final BoxRenderCalculator calculator;
    private final BoxStyle style;

    private ShapeRenderer boxRenderer;
    private Rectangle box;

    private ShapeRenderer borderRenderer;
    private Rectangle border;

    @AssistedInject
    public BoxComponent(@Assisted BoxStyle style) {
        calculator = new BoxRenderCalculator(style);
        this.style = style;
    }

    @Override
    public void init(ScreenState state) {
        boxRenderer = new ShapeRenderer();
        boxRenderer.setAutoShapeType(true);
        boxRenderer.setColor(style.getBoxColor());
        box = new Rectangle();

        borderRenderer = new ShapeRenderer();
        borderRenderer.setAutoShapeType(true);
        borderRenderer.setColor(style.getBorderColor());
        border = new Rectangle();
    }

    @Override
    public void render(float delta, ScreenState state) {
        Gdx.gl.glEnable(GL20.GL_BLEND);
        Gdx.gl.glBlendFunc(GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA);
        boxRenderer.begin(ShapeRenderer.ShapeType.Filled);
        boxRenderer.rect(calculator.getBoxX(), calculator.getBoxY(), calculator.getBoxWidth(), calculator.getBoxHeight());
        boxRenderer.end();

        if (!style.isUseBorder()) {
            return;
        }

        List<Color> borderGradient = GradientResource.gradients.get(style.getBorderGradient());
        Gdx.gl.glLineWidth(calculator.getBorderThickness());
        borderRenderer.begin(ShapeRenderer.ShapeType.Point);
        if (borderGradient != null) {
            borderRenderer.rect(
                    calculator.getBorderX(), calculator.getBorderY(),
                    calculator.getBorderWidth(), calculator.getBorderHeight(),
                    borderGradient.get(0), borderGradient.get(1), borderGradient.get(2), borderGradient.get(3)
            );
        } else {
            borderRenderer.rect(
                    calculator.getBorderX(), calculator.getBorderY(),
                    calculator.getBorderWidth(), calculator.getBorderHeight()
            );
        }
        borderRenderer.end();
        Gdx.gl.glDisable(GL20.GL_BLEND);
    }

    @Override
    public void destroy() {
        boxRenderer.dispose();
        borderRenderer.dispose();
    }

    public void setPosition(float x, float y) {
        calculator.updatePosition(x, y);
    }
}

File: core/src/com/alta_v2/rendering/common/component/box/BoxComponent.kt
package com.alta_v2.rendering.common.component.box

import com.alta_v2.rendering.Renderer
import com.alta_v2.rendering.ScreenState
import com.alta_v2.rendering.common.utils.gradients
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.google.inject.assistedinject.Assisted
import com.google.inject.assistedinject.AssistedInject

class BoxComponent @AssistedInject constructor(@Assisted private val style: BoxStyle) : Renderer {

    private lateinit var boxRenderer: ShapeRenderer
    private lateinit var borderRenderer: ShapeRenderer
    private val calculator = BoxRenderCalculator(style)

    override fun init(state: ScreenState) {
        boxRenderer = ShapeRenderer()
        boxRenderer.setAutoShapeType(true)
        boxRenderer.color = style.boxColor

        borderRenderer = ShapeRenderer()
        borderRenderer.setAutoShapeType(true)
        borderRenderer.color = style.borderColor
    }

    override fun render(delta: Float, state: ScreenState) {
        Gdx.gl.glEnable(GL20.GL_BLEND)
        Gdx.gl.glBlendFunc(GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA)

        boxRenderer.begin(ShapeRenderer.ShapeType.Filled)
        boxRenderer.rect(calculator.boxX, calculator.boxY, calculator.boxWidth.toFloat(), calculator.boxHeight.toFloat())
        boxRenderer.end()
        if (style.useBorder) {
            renderBorder()
        }
    }

    override fun destroy() {
        boxRenderer.dispose()
        borderRenderer.dispose()
    }

    fun setPosition(x: Float, y: Float) {
        calculator.updatePosition(x, y)
    }

    private fun renderBorder() {
        val borderGradient = gradients[style.borderGradient]
        Gdx.gl.glLineWidth(calculator.borderThickness.toFloat())
        borderRenderer.begin(ShapeRenderer.ShapeType.Point)
        if (borderGradient != null) {
            borderRenderer.rect(
                    calculator.borderX, calculator.borderY,
                    calculator.borderWidth.toFloat(), calculator.borderHeight.toFloat(),
                    borderGradient[0], borderGradient[1], borderGradient[2], borderGradient[3]
            )
        } else {
            borderRenderer.rect(
                    calculator.borderX, calculator.borderY,
                    calculator.borderWidth.toFloat(), calculator.borderHeight
                    .toFloat())
        }
        borderRenderer.end()
        Gdx.gl.glDisable(GL20.GL_BLEND)
    }
}